framework/src/utils/RankMap.C

```diff
@@ -0,0 +1,77 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RankMap.h"
+
+#include "PerfGraphInterface.h"
+
+RankMap::RankMap(const Parallel::Communicator & comm, PerfGraph & perf_graph)
+  : ParallelObject(comm),
+    PerfGraphInterface(perf_graph, "RankMap"),
+    _construct_timer(registerTimedSection("construct", 2))
+{
+  TIME_SECTION(_construct_timer);
+
+  auto num_procs = n_processors();
+
+  _rank_to_hardware_id.resize(num_procs);
+
+  // This will be the world rank of the root process
+  // from the shared memory communicator we're getting ready to create
+  // Each process on the same node will end up with the same world_rank
+  processor_id_type world_rank = 0;
+
+#ifdef LIBMESH_HAVE_MPI
```

Comment:
Splitting is supported in the libMesh communicator object. You can drop this ifdef and and the raw MPI here.

Reply:
OK - I opened a ticket so we can eventually pull this back out:
https://github.com/libMesh/libmesh/issues/2128